content/blog/_posts/2021-04-rust-custom-test-harness.md

```diff
@@ -0,0 +1,296 @@
+
+This is distinctive, because libtest will compile each of your `#[test]` labeled functions into its own binary crate (with its own `main()`) and executes it as part of the test. 
+
+First we’re going to create an integration test directory at the root of the crate where we’re going to build our integration test focused binary.
+
+```shell
+$ mkdir integration
+$ touch integration/main.rs
+
+# Then create a main() function in main.rs
+```
+
+In your Cargo.toml, you want to add 
+
+```toml
+# Cargo.toml
+
+[dev-dependencies]
+inventory = "0.1"
+
+[[test]]
+name = “integration”
+path = “integration/main.rs”
+harness = false
+```
+
+What this does is tell cargo test to not use libtest when running the `integration` test.
+
+Then when we run `cargo test integration`, what cargo will do is compile `integration/main.rs` and execute it in the same manner as `cargo run`. This is all a harness is from `cargo`’s perspective. 
+
+(We'll revisit the `inventory` crate later in the post)
+
+
+### Add Setup and teardown steps
+
+First thing we’ll do is lay the foundation for our testing pattern. We’ll create 2 functions, `setup()` and `teardown()`, and add them to our `main()` (with reserved space inbetween for our future tests to be called).
+
+```rust
+// main.rs
+
+fn setup() {
+   println!("Setup")
+}
+
+fn teardown() {
+   println!("Teardown")
+}
+fn main() {
+   // Setup test environment
+   setup();
+
+   // TODO: Run the test
+
+   // Teardown test environment
+   teardown();
+}
+```
+
+### Collect all integration tests
+
+What do I mean? Just use `#[test]`, right? That’s what I was thinking at least. Turns out that is you opt out of using `#[test]` with `harness = false`. 
+
+
+#### How does libtest collect tests?
+
+After digging around in the [libtest](https://github.com/rust-lang/libtest/blob/master/libtest/lib.rs) and [Cargo](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_test.rs) and [Rust](https://github.com/rust-lang/rust/blob/master/compiler/rustc_builtin_macros/src/test.rs) code, the short answer is: it isn’t clear and it isn't reusable.
+
+If that surprises you, then you were like me. I was hoping to use the test collection functionality from `#[test]`, but it wasn’t clear how to accomplish this. My mental model for how `cargo test` works needed a refresh.
+
+This gives you 2 practical options for collecting tests
+
+1. Manually modify `integration/main.rs` and add your test in between the setup and teardown
+    * A quick and straightforward solution if you have a small set of tests
+2. Build a test collector. We generate an external test catalog, and modify `integration/main.rs` to execute tests from the catalog.
+    * This is a long term solution, which we’ll be covering for the rest of the post.
+
+### Building the test collector
+
+The [inventory](https://crates.io/crates/inventory) crate markets itself as a plugin registry system. It will be doing all the heavy lifting. In this case, our tests are the plugins. 
```

Comment:
I'd avoid saying "markets", it doesn't feel right in this context. Maybe say something like

> The [inventory] crate is a plugin registry system. We'll be using it for all the heavy lifting in our test framework - we'll be treating our tests as the plugins